crates/red_knot_python_semantic/resources/mdtest/known_constants.md

```diff
@@ -0,0 +1,41 @@
+# Known constants
+
+## `typing.TYPE_CHECKING`
+
+This constant is `True` when in type-checking mode, `False` otherwise. The symbol is defined to be
+`False` at runtime. In typeshed, it is annotated as `bool`. This tests makes sure that we infer
+`Literal[True]` for it anyways.
+
+### Basic
+
+```py
+from typing import TYPE_CHECKING
+import typing
+
+reveal_type(TYPE_CHECKING)  # revealed: Literal[True]
+reveal_type(typing.TYPE_CHECKING)  # revealed: Literal[True]
+```
+
+### Aliased
+
+Make sure that we still infer the correct type if the constant has been given a different name:
+
+```py
+from typing import TYPE_CHECKING as TC
+
+reveal_type(TC)  # revealed: Literal[True]
+```
+
+### Only from `typing`
+
+Make sure we only use our special handling for `typing.TYPE_CHECKING` and not for other modules:
+
+```py path=constants.py
+TYPE_CHECKING: bool = False
+```
+
+```py
+from constants import TYPE_CHECKING
+
+reveal_type(TYPE_CHECKING)  # revealed: bool
+```
```

Comment:
Hrm... not sure... I feel like I'd prefer for users to explicitly opt into something like that using something like mypy's [`--always-true` option](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-always-true). Pyright also has a similar option. It has the advantage that it's more general, more explicit, more configurable and less magical.

Reply:
Looks like pyright's version of this configuration knob is the `defineConstant` setting: https://github.com/microsoft/pyright/blob/main/docs/configuration.md#environment-options

Reply:
Hmm, I'm curious where else (and why else) people use this feature. But apparently people do, if pyright and mypy both support it?

I'm fine with leaving this as a separate feature to be handled later. According to GH code search, it seems like there's a reasonable amount of `TYPE_CHECKING = False` out there; I won't be surprised if we end up feeling like we need to support it by default.

Reply:
I know that people who use both mypy and pyright often use it for cross-type-checker compatibility. You can configure mypy with `--always-true=MYPY` and pyright with `defineConstant = {"MYPY": False}`, and then do things like this so that mypy will only see the first branch and pyright will only see the second:

```py
if MYPY:
    ...
else:
    ...
```

Reply:
I think the test is fine as-is for now; we can add the known-constants feature separately, and later evaluate if we feel we need to follow mypy and pyright in making `TYPE_CHECKING` a known-constant by default.